### PR TITLE
add neural network modules that work with tensor maps

### PR DESCRIPTION
**Pull request by agoscinski**
From the discussion in #13. Should replace https://github.com/bananenpampe/H2O/blob/move-rascaline/model/nn/modules.py

The goal is to offer a flexible wrapper class around torch modules that are applied on each block of a input TensorMap while also offer functions to wrap simpler cases, for example when the same module is used for all blocks (for that there is the `from_module` constructor).

What is missing, is a functionality that allows to merge multiple blocks into a new one (e.g. in species compression), but I also don't think it is a good idea to put this in the same class.

TODO (will add these after first review):
- [x] add another test for using the `__init__` of `Linear`
- [x] add another test for using the `from_module` of `ModuleTensorMap`
- [x] move `equisolve.nn` to `equisolve.torch.nn` (needs to be discussed in meeting) 
- [x] rascaline has been updated https://github.com/Luthaf/rascaline/pull/232 we can remove the requirements.txt (maybe keeping it is smarter because of cross linking of packages with sphinx, need to check this.)
----
:books: Documentation preview :books:: https://equisolve--62.org.readthedocs.build/en/62/

### CODE REVIEW

**Comment by PicoCentauri** [comment]
> What is missing, is a functionality that allows to merge multiple blocks into a new one (e.g. in species compression), but I also don't think it is a good idea to put this in the same class.

This can be done by the user using vanilla tensormap functions, right?

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -28,8 +28,14 @@ jobs:
...
-    - name: run Python tests
-      run: tox -e tests
+    - name: run Python metatensor-core tests
+      run: tox -e metatensor-core-numpy-tests
+
+    - name: run Python metatensor-core tests
+      run: tox -e metatensor-core-torch-tests
+
+    - name: run Python metatensor-torch tests
+      run: tox -e metatensor-torch-tests
```
Comment (PicoCentauri) [inline .github/workflows/tests.yml:38]: I think you can remove the `metatensor` from the name. We don't have tests without metatensor I assume.

File: docs/requirements.txt [hunk 2]
```diff
@@ -10,3 +10,7 @@ furo
 rascaline @ https://github.com/luthaf/rascaline/archive/fb5332f.zip
 sphinx >=4.4
 sphinx-gallery
+# temporary add torch to dependencies until rascaline is on metatensor
+# version 3c5fee40
```
Comment (PicoCentauri) [inline docs/requirements.txt:14]:
```suggestion
# version 3c5fee40
```

File: src/equisolve/nn/__init__.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+try:
+    import torch  # noqa: F401
+
+    HAS_TORCH = True
+except ImportError:
+    HAS_TORCH = False
+
+if HAS_TORCH:
+    from .module_tensor import LinearTensorMap, ModuleTensorMap  # noqa: F401
```
Comment (Luthaf) [inline src/equisolve/nn/__init__.py:9]: naming wise, I would call these `Linear` and `Module`, and users can then use `torch.nn.Linear/torch.nn.Module` or `equisolve.nn.Linear/equisolve.nn.Module` for the different behavior.

**Review by Luthaf — COMMENTED** [review]
The code looks good, the documentation will still need some work: how to wrap custom module inside `equisolve.nn.Module`, how/when to use these/... but this can come at a later date.

File: src/equisolve/nn/module_tensor.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                    )
+                out_block.add_gradient(
+                    parameter=parameter,
+                    gradient=self.forward_block(key, gradient),
+                )
+            out_blocks.append(out_block)
+
+        return TensorMap(tensor.keys, out_blocks)
+
+    def forward_block(self, key: LabelsEntry, block: TensorBlock) -> TensorBlock:
```
Comment (PicoCentauri) [inline src/equisolve/nn/module_tensor.py:94]: I think this can also be a Labels object right?

File: src/equisolve/nn/module_tensor.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    """
+    ModuleDictTensorMap is needed when the keys correspond to different features sizes
+
+    :param module_map: a dictionary of modules with tensor map keys as dict keys
+        each module is applied on a block
+
+    :param out_tensor: a tensor map that has the output properties of the Labels
+    """
+
+    def __init__(self, module_map: ModuleDict, out_tensor: Optional[TensorMap] = None):
```
Comment (PicoCentauri) [inline src/equisolve/nn/module_tensor.py:47]: I am not sure what the `out_tensor` is doing. Is this the `TensorMap` where the results will be written in?

File: src/equisolve/nn/module_tensor.py (hunk 4 above)
Comment (Luthaf) [reply src/equisolve/nn/module_tensor.py:94]: if you get it by iterating over the keys, you'll get `LabelsEntry`

**Review by agoscinski — COMMENTED** [review]
> > What is missing, is a functionality that allows to merge multiple blocks into a new one (e.g. in species compression), but I also don't think it is a good idea to put this in the same class.
>
> This can be done by the user using vanilla tensormap functions, right?

Yes, I am not sure if one can create another Module for this purpose that is generic enough to be useful for multiple applications without being too abstract and less efficient than doing it "manually". For example the species decomposition in alchemical learning. You are transforming the species neighbor sparsity to a new property the pseudo species. So this combines a movement to properties and a linear combination. Don't have good ideas here.

---

> The code looks good, the documentation will still need some work: how to wrap custom module inside equisolve.nn.Module, how/when to use these/... but this can come at a later date.

Improved the documentation. I will do an issue for an example for the neural network.

**Review by PicoCentauri — APPROVED** [review]
Thanks for the improced Docs @agoscinski !
